File: AutoTestJava/AutoTest/src/test/java/ru/auto/BasePage.java
package ru.auto;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;

import java.time.Duration;

public class BasePage {
    protected WebDriver driver;

    public BasePage(WebDriver driver) {
        this.driver = driver;
    }

    protected WebElement find(By locator) {
        return driver.findElement(locator);
    }

    protected void click(By locator, int timeout) {
        waitUntilElementIsVisible(locator, timeout);
        find(locator).click();
    }

    protected void type(By locator, String text, int timeout) {
        waitUntilElementIsVisible(locator, timeout);
        find(locator).sendKeys(text);
    }

    protected void waitUntilElementIsVisible(By locator, int timeout) {
        WebDriverWait wait = new WebDriverWait(driver, Duration.ofSeconds(timeout));
        wait.until(ExpectedConditions.visibilityOfElementLocated(locator));
    }

    protected void openUrl(String url) {
        driver.get(url);
    }

    protected String getText(By locator, int timeout) {
        waitUntilElementIsVisible(locator, timeout);
        return find(locator).getText();
    }
}

File: Лабораторная работа 5/AutoTest/base_page.py
from selenium.common import NoSuchElementException
from selenium.webdriver.remote.webdriver import WebDriver
from selenium.webdriver.remote.webelement import WebElement
from selenium.webdriver.support.wait import WebDriverWait
from selenium.webdriver.support import expected_conditions as ec

class BasePage:
    # Драйвер для работы с веб-страницами
    def __init__(self, driver: WebDriver):
        self._driver = driver

    def _scroll_into_view(self, locator: tuple):
        element = self._find(locator)
        self._driver.execute_script("arguments[0].scrollIntoView(true);", element)

    # Поиск и возврат одного элемента на веб-странице с помощью переданного локатора
    def _find(self, locator: tuple) -> WebElement:
        return self._driver.find_element(*locator)
    
    # Ввод текста в элемент на веб-странице
    def _type(self, locator: tuple, text: str, time: int = 10):
        # Прокручиваем элемент в видимую область
        self._scroll_into_view(locator)
        self._wait_until_element_is_visible(locator, time)
        self._find(locator).send_keys(text)

            # выполнение клика по элементу на веб-странице
    def _click(self, locator: tuple, time: int = 10):
        self._wait_until_element_is_visible(locator, time)
        self._find(locator).click()

    # ожидание появления элемента на веб-странице до тех пор, пока он не станет видимым
    def _wait_until_element_is_visible(self, locator: tuple, time: int = 10):
        wait = WebDriverWait(self._driver, time)
        wait.until(ec.visibility_of_element_located(locator))

    # открытие указанного URL веб-страницы
    def _open_url(self, url: str):
        self._driver.get(url)

    # получение текстового содержимого элемента, найденного по указанному локатору.
    def _get_text(self, locator: tuple, time: int = 3) -> str:
        self._wait_until_element_is_visible(locator, time)
        return self._find(locator).text.encode("utf-8").decode("utf-8")
